zebrad/src/components/mempool/crawler.rs

```diff
@@ -39,26 +46,46 @@ where
     S::Future: Send,
 {
     /// Spawn an asynchronous task to run the mempool crawler.
-    pub fn spawn(peer_set: S) -> JoinHandle<()> {
+    pub fn spawn(peer_set: S, latest_sync_length: watch::Receiver<Vec<usize>>) -> JoinHandle<()> {
         let crawler = Crawler {
             peer_set: Mutex::new(Timeout::new(peer_set, PEER_RESPONSE_TIMEOUT)),
+            latest_sync_length,
         };
 
         tokio::spawn(crawler.run())
     }
 
     /// Periodically crawl peers for transactions to include in the mempool.
-    pub async fn run(self) {
-        loop {
-            self.wait_until_enabled().await;
+    pub async fn run(mut self) {
+        while self.wait_until_enabled().await {
             self.crawl_transactions().await;
             sleep(RATE_LIMIT_DELAY).await;
         }
     }
 
     /// Wait until the mempool is enabled.
-    async fn wait_until_enabled(&self) {
-        // TODO: Check if synchronizing up to chain tip has finished (#2603).
+    ///
+    /// Returns `false` if there's no way to know if the mempool is enabled and the crawler should
+    /// stop.
+    async fn wait_until_enabled(&mut self) -> bool {
+        while !self.is_enabled() {
+            let wait_result = self.latest_sync_length.changed().await;
+
+            if wait_result.is_err() {
+                // If this happens, it's likely that Zebra is shutting down.
+                debug!("Mempool crawler stopped receiving latest sync lengths. Stopping...");
+                return false;
+            }
+        }
+
+        true
+    }
+
+    /// Check if the mempool is currently enabled.
+    fn is_enabled(&self) -> bool {
+        let sync_lengths = self.latest_sync_length.borrow();
+
+        is_enabled(&*sync_lengths)
     }
```

Comment:
This looks like a good interface for activating mempool tasks and services.
Having both a sync and `async` interface will be very useful.

I think we'll want to have similar methods in at least 4 mempool components:
https://jamboard.google.com/d/1HUhdqx8afJUqgiBdBcjCw0aAfDod9R9KBJeAH24U1Qg/viewer

Can we move `latest_sync_length`, `wait_until_enabled`, and `is_enabled` into a wrapper type?
That way, we'll have a single implementation, which will be easier to use and modify.

Can you put the wrapper type in `sync::recent_sync_lengths`?
We'll end up using it for both the mempool and the state block commit advertisements.

Reply:
Merged `SyncStatus` in #2685.

So I guess we want to rewrite this PR to use `SyncStatus` now?

Reply:
Done.